### PR TITLE
Concurrency: Hop back to the previous executor after actor calls.

### PR DESCRIPTION
**Pull request by jckarter**
Tasks shouldn't normally hog the actor context indefinitely after making a call that's bound to
that actor, since that prevents the actor from potentially taking on other jobs it needs to
be able to address. Set up SILGen so that it saves the current executor (using a new runtime
entry point) and hops back to it after every actor call, not only ones where the caller context
is also actor-bound.

The added executor hopping here also exposed a bug in the runtime implementation while processing
DefaultActor jobs, where if an actor job returned to the processing loop having already yielded
the thread back to a generic executor, we would still attempt to make the actor give up the thread
again, corrupting its state. This is fixed as well.

rdar://71905765

### CODE REVIEW

**Comment by jckarter** [comment]
@swift-ci Please test

File: lib/SILGen/SILGenLValue.cpp [hunk 1]
```diff
@@ -1588,7 +1589,7 @@ namespace {
       assert(getAccessorDecl()->isGetter());
 
       SILDeclRef getter = Accessor;
-      bool didHop = false;
+      ExecutorBreadcrumb didHop;
```
Comment (etcwilde) [inline lib/SILGen/SILGenLValue.cpp:1592]: `didHop` made sense as a boolean, but doesn't seem clear in the current use. I see that below, in `SILGenFunction::emitHopToTargetActor`, you're just calling it `breadcrumb`. Can we change the name to more accurately describe the use?

**Review by kavon — COMMENTED** [review]
This looks great! One thing though: there's another non-isolated async function, `accessGlobalIsolatedMember`, in the `hop_to_executor_async_prop.swift` regression test that did not catch these new changes, but should be updated now to expect the hop-back. I can't select the lines in GitHub review for some reason, but it's lines 119-128 in that file as of commenting.

**Comment by jckarter** [comment]
I also had to add TSAN edges to all the places actors transition in and out of running state, so that we don't get spurious data race errors when an actor is accessed by a different thread after we've transitioned a task away from its executor.

**Comment by jckarter** [comment]
@swift-ci Please test

**Comment by jckarter** [comment]
@swift-ci Please test Linux

**Pull request merged**